src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java

```diff
@@ -0,0 +1,105 @@
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+public class MaintenanceFilter implements Filter {
+
+    private static final String MAINTENANCE_URI = "/maintenance";
+    private static final String MAIN_URI = "/";
+    private static final String OAUTH_PATH = "/oauth";
+    private final boolean maintenanceEnabled;
+    private final String maintenanceOverrideTokenName;
+    private final String maintenanceOverrideTokenValue;
+    private static final String ASSETS_PATH = "/assets";
+
+    @Autowired
+    public MaintenanceFilter(@Value("${maintenance.enabled}") boolean maintenanceEnabled,
+                             @Value("${maintenance.overrideTokenName}") String maintenanceOverrideTokenName,
+                             @Value("${maintenance.overrideTokenValue}") String maintenanceOverrideTokenValue) {
+        this.maintenanceEnabled = maintenanceEnabled;
+        this.maintenanceOverrideTokenName = maintenanceOverrideTokenName;
+        this.maintenanceOverrideTokenValue = maintenanceOverrideTokenValue;
+
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletResponse httpResponse = (HttpServletResponse) response;
+        if (shouldRedirectToMaintenancePage(request)) {
+            httpResponse.sendRedirect(MAINTENANCE_URI);
+            return;
+        }
+
+        if (shouldRedirectToMainPage(request)) {
+            httpResponse.sendRedirect(MAIN_URI);
+            return;
+        }
+
+        chain.doFilter(request, response);
+    }
+
+
+    @Override
+    public void destroy() {
+
+    }
+
+    private boolean shouldRedirectToMaintenancePage(ServletRequest request) {
+        HttpServletRequest httpRequest = (HttpServletRequest)request;
+        return maintenanceEnabled &&
+                !isMaintenaceRequest(httpRequest) &&
+                !isWhiteListedRequest(httpRequest) &&
+                !isMaintenanceOverrideCookiePresent(httpRequest);
+    }
+
+    private boolean shouldRedirectToMainPage(ServletRequest request) {
+        HttpServletRequest httpRequest = (HttpServletRequest)request;
+        return !maintenanceEnabled &&
+                isMaintenaceRequest(httpRequest);
+    }
+
+    private boolean isMaintenaceRequest(HttpServletRequest httpRequest) {
+        return httpRequest.getRequestURI().equalsIgnoreCase(MAINTENANCE_URI);
+    }
+
+    private boolean isWhiteListedRequest(HttpServletRequest httpRequest) {
+        return httpRequest.getRequestURI().startsWith(ASSETS_PATH) ||
```

Comment:
why we whitelist `/assets` path?

Reply:
as the filter handles all requests - hope that explains it?

Reply:
as per the settings in `UserSecurityConfig` 
`.antMatchers("/login", "/oauth/logout", "/webjars/**", "/assets/**", "/signup/**", "/reset/**",  "/account/passwordUpdated").permitAll()`
why paths like /reset or /webjars are not whitelisted?

Reply:
cause we only need /assets/* and /oauth/* for the override to work. It is a whitelist for the maintenance mode override not whitelist authentication/authorisation wise :)